File: app/src/main/java/ch/unstable/ost/preference/StationDaoLoader.java
package ch.unstable.ost.preference;


import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;

import ch.unstable.ost.R;
import ch.unstable.ost.api.StationsDAO;

public class StationDaoLoader {
    public interface StationDAOFactory {
        @NonNull
        StationsDAO getStationsDAO(@NonNull Context context);
    }

    @NonNull
    public static StationsDAO createStationDAO(final Context context) {
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        String preference = sharedPreferences.getString(PreferenceKeys.KEY_STATIONS_DAO, context.getString(R.string.prefs_station_dao_factory_default));

        try {
            Class<?> clazz = Class.forName(preference);
            StationDAOFactory daoFactory = (StationDAOFactory) clazz.newInstance();
            return daoFactory.getStationsDAO(context);
        } catch (Exception e) {
            throw new IllegalStateException("Couldn't get factory for StationsDAO", e);
        }
    }
}

File: app/src/main/java/ch/unstable/ost/preference/StationDaoLoader.kt
package ch.unstable.ost.preference

import android.content.Context
import android.preference.PreferenceManager
import ch.unstable.ost.R
import ch.unstable.ost.api.StationsDAO

object StationDaoLoader {
    fun createStationDAO(context: Context): StationsDAO {
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
        val className = sharedPreferences.getString(PreferenceKeys.KEY_STATIONS_DAO, context.getString(R.string.prefs_station_dao_factory_default))!!
        return try {
            val clazz = Class.forName(className)
            val daoFactory = clazz.getDeclaredConstructor().newInstance() as StationDAOFactory
            daoFactory.getStationsDAO(context)
        } catch (e: Exception) {
            throw IllegalStateException("Couldn't get factory for StationsDAO $className", e)
        }
    }

    interface StationDAOFactory {
        fun getStationsDAO(context: Context): StationsDAO
    }
}